### PR TITLE
feat(planning_test_utils): add off-track test for path and trajectory

### PR DESCRIPTION
**Pull request by TakaHoribe**
## Description

Add util functions to check the node will not die when the ego-vehicle pose is located far from the target trajectory.

## Related links

None

## Tests performed

None. (tested in each package)

## Interface changes

None

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/planning_test_utils/include/planning_interface_test_manager/planning_interface_test_manager_utils.hpp [hunk 1]
```diff
@@ -235,10 +235,10 @@ HADMapBin makeMapBinMsg()
   return map_bin_msg;
 }
 
-Odometry makeOdometry()
+Odometry makeOdometry(const double shift = 0.0)
 {
   Odometry odometry;
-  const std::array<double, 4> start_pose{5.5, 4., M_PI_2};
+  const std::array<double, 4> start_pose{0.0, shift, 0.0, 0.0};
```
Comment (TakaHoribe) [inline planning/planning_test_utils/include/planning_interface_test_manager/planning_interface_test_manager_utils.hpp:241]: @kyoichi-sugahara Why the pose was defined as follows?
```
const std::array<double, 4> start_pose{5.5, 4., M_PI_2};
```

Comment (kyoichi-sugahara) [reply planning/planning_test_utils/include/planning_interface_test_manager/planning_interface_test_manager_utils.hpp:241]: @TakaHoribe 
I copied from [here](https://github.com/autowarefoundation/autoware.universe/blob/c0ebaf351103c8343021b3d1f7e013b74462cfae/planning/freespace_planning_algorithms/test/src/test_freespace_planning_algorithms.cpp/#L42) to do similar test.
But in the interface test for freespace_planer, pose doesn't need to be defined with `const std::array<double, 4> start_pose{5.5, 4., M_PI_2};` and agree with your difinition.

**Comment by kyoichi-sugahara** [comment]
test success with following command
```
colcon test --event-handlers console_cohesion+ --packages-select behavior_path_planner planning_validator motion_velocity_smoother motion_velocity_smoother obstacle_stop_planner obstacle_stop_planner obstacle_stop_planner freespace_planner
```

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM and test success with new behavior_path_planner architecture.

File: planning/planning_test_utils/src/planning_interface_test_manager.cpp [hunk 2]
```diff
@@ -401,14 +401,24 @@ void PlanningInterfaceTestManager::testOffTrackFromPath(rclcpp::Node::SharedPtr
...
 void PlanningInterfaceTestManager::testOffTrackFromTrajectory(rclcpp::Node::SharedPtr target_node)
 {
-  // write me
-  (void)target_node;
+  publishNominalTrajectory(target_node, input_trajectory_name_);
+
+  const std::vector<double> deviation_from_traj = {0.0, 1.0, 10.0, 100.0};
+  for (const auto & deviation : deviation_from_traj) {
+    publishOdometry(target_node, input_odometry_name_, deviation);
+    test_utils::spinSomeNodes(test_node_, target_node, 5);
```
Comment (kyoichi-sugahara) [inline planning/planning_test_utils/src/planning_interface_test_manager.cpp:420]: spinSomeNodes is executed in the publish odometry function [here](https://github.com/autowarefoundation/autoware.universe/blob/c0ebaf351103c8343021b3d1f7e013b74462cfae/planning/planning_test_utils/include/planning_interface_test_manager/planning_interface_test_manager_utils.hpp#LL368C3-L368C13)
so `spinSomeNodes` here is not necessary

**Review by kyoichi-sugahara — CHANGES_REQUESTED** [review]
Please check duplication of the `spinSomeNodes` function

File: planning/planning_test_utils/include/planning_interface_test_manager/planning_interface_test_manager_utils.hpp (hunk 1 above)
Comment (TakaHoribe) [reply planning/planning_test_utils/include/planning_interface_test_manager/planning_interface_test_manager_utils.hpp:241]: Thanks! done.

File: planning/planning_test_utils/src/planning_interface_test_manager.cpp (hunk 2 above)
Comment (TakaHoribe) [reply planning/planning_test_utils/src/planning_interface_test_manager.cpp:420]: @kyoichi-sugahara In both `publishOdometry()` and `publishNominalTrajectory()`, spin is called only once. Considering the possibility that a topic may be dropped or may not be published due to topics stored in the queue before, it is better to call spin for sometimes. 
Actually, the topic drop often happens in the trajectory_follower_node testing.

Comment (TakaHoribe) [reply planning/planning_test_utils/src/planning_interface_test_manager.cpp]: Addressed here: https://github.com/autowarefoundation/autoware.universe/pull/3587/commits/3fd8e01f65802f6d0e72b9770970885541cb25b5

File: planning/planning_test_utils/include/planning_interface_test_manager/planning_interface_test_manager_utils.hpp [hunk 3]
```diff
@@ -353,14 +353,15 @@ void spinSomeNodes(
 template <typename T>
 void publishToTargetNode(
   rclcpp::Node::SharedPtr test_node, rclcpp::Node::SharedPtr target_node, std::string topic_name,
-  typename rclcpp::Publisher<T>::SharedPtr publisher, T data, const int repeat_count = 1)
+  typename rclcpp::Publisher<T>::SharedPtr publisher, T data, const int repeat_count = 3)
```
Comment (kyoichi-sugahara) [inline planning/planning_test_utils/include/planning_interface_test_manager/planning_interface_test_manager_utils.hpp:356]: agree with this change!

**Review by kyoichi-sugahara — APPROVED** [review]

**Pull request merged**